### ISSUE TITLE
[Question] Reproducibility in OneShotAcquisitionFunction optimization

### ISSUE THREAD

**Issue by LLipschitz** [pytorch/botorch#1190]
Given that the `OneShotAcquisitionFunction` optimization is a deterministic optimization problem, why does running `optimize_acqf` on it return a different (typically very drastic) candidate each time we optimize? If I can set a random seed for the restart candidates and the optimization, where would that be in the code? Thanks!

Edit: Looks like passing `options={"seed":seed}` to the initial samples generator will enable reproducibility, based on a quick trial.

**Comment by saitcakmak** [comment]
Once constructed, an instance of the acquisition function should be deterministic. You can verify this by evaluating it repeatedly with the same input (call `acqf(X)` etc).

As you found out, passing the seed option will do it. It sets the seed for the `SobolEngine` under the hood, which is used to generate the initial candidates. Alternatively, you could set the torch seed before calling `optimize_acqf` via `torch.manual_seed(seed)`, or you could pass manually generated `batch_initial_conditions`.

**Comment by Balandat** [comment]
To add to that - the reason why the suggested candidates are different even though the acquisition function is deterministic is b/c this is a very hard nonconvex optimization problem so we can’t expect to reliably find the global optimum. Fixing the seed will generate the same initial conditions for the random restart optimization and this should result in a reproducible solution.

**Comment by LLipschitz** [comment]
Thanks. Just a one comment. Setting `torch.manual_seed(seed)` before acquisition function optimization did not ensure reproducibility. 
Closing issue.

**Comment by saitcakmak** [comment]
> Setting torch.manual_seed(seed) before acquisition function optimization did not ensure reproducibility.

Interesting. Did you use the same acquisition function object? It works for me:

```
import torch
from botorch.acquisition import qKnowledgeGradient
from botorch.fit import fit_gpytorch_model
from botorch.models import SingleTaskGP
from botorch.models.transforms import Standardize
from botorch.optim import optimize_acqf
from botorch.test_functions import Ackley
from botorch.utils.transforms import unnormalize
from gpytorch.mlls import ExactMarginalLogLikelihood


func = Ackley(negate=True)
X = torch.rand(20, 2)
Y = func(unnormalize(X, func.bounds)).view(-1, 1)
model = SingleTaskGP(X, Y, outcome_transform=Standardize(m=1))
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(mll)
acqf = qKnowledgeGradient(model)

bounds = torch.zeros(2, 2)
bounds[1] = 1

for i in range(10):
    torch.manual_seed(0)
    print(
        optimize_acqf(
            acq_function=acqf,
            bounds=bounds,
            q=1,
            num_restarts=2,
            raw_samples=4,
        )
    )
```

output:
```

/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
/mnt/xarfuse/uid-352651/ebf4dbc9-seed-nspid4026531836_cgpid32757103-ns-4026531840/botorch/models/gpytorch.py:88: BotorchTensorDimensionWarning:

Non-strict enforcement of botorch tensor conventions. Ensure that target tensors Y has an explicit output dimension.

(tensor([[0.6602, 0.5207]]), tensor(-18.6861))
```
